File: apps/api/app/api_client/endpoints/certificates.py
"""API client for certificate endpoints with typed responses."""

from app.api_client.client import api_client
from api.schemas.certificate import CertificateCreate, CertificateResponse, CertificateUpdate


class CertificatesAPI:
    """API client for certificate endpoints with typed responses."""

    @staticmethod
    async def list_certificates(user_id: int) -> list[CertificateResponse]:
        """List all certificates for a user. Returns list of CertificateResponse models."""
        return await api_client.get(
            "/api/v1/certificates",
            params={"user_id": user_id},
            response_model=CertificateResponse,
        )

    @staticmethod
    async def create_certificate(
        user_id: int,
        title: str,
        date: str,
        institution: str | None = None,
    ) -> CertificateResponse:
        """Create a new certificate. Returns CertificateResponse model."""
        return await api_client.post(
            "/api/v1/certificates",
            params={"user_id": user_id},
            json=CertificateCreate(
                title=title,
                institution=institution,
                date=date,
            ).model_dump(),
            response_model=CertificateResponse,
        )

    @staticmethod
    async def update_certificate(
        certificate_id: int,
        title: str | None = None,
        institution: str | None = None,
        date: str | None = None,
    ) -> CertificateResponse:
        """Update a certificate. Returns CertificateResponse model."""
        update_data = CertificateUpdate(
            title=title,
            institution=institution,
            date=date,
        ).model_dump(exclude_unset=True)
        return await api_client.patch(
            f"/api/v1/certificates/{certificate_id}",
            json=update_data,
            response_model=CertificateResponse,
        )

    @staticmethod
    async def delete_certificate(certificate_id: int) -> None:
        """Delete a certificate."""
        await api_client.delete(f"/api/v1/certificates/{certificate_id}")

File: apps/web/src/lib/api/certificates.ts
/**
 * Certifications API client endpoints.
 */

import { apiRequest } from "./client";
import type { Certification, CertificationInsert, CertificationUpdate } from "@resume/database/types";

export const certificatesAPI = {
  /**
   * List all certifications for a user.
   */
  list: (userId: number): Promise<Certification[]> => {
    const queryParams = new URLSearchParams({
      user_id: userId.toString(),
    });
    return apiRequest<Certification[]>(`/api/certifications?${queryParams.toString()}`);
  },

  /**
   * Get a specific certification.
   */
  get: (certificationId: number): Promise<Certification> => {
    return apiRequest<Certification>(`/api/certifications/${certificationId}`);
  },

  /**
   * Create a new certification.
   */
  create: (params: { user_id: number; data: Omit<CertificationInsert, "user_id"> }): Promise<Certification> => {
    const queryParams = new URLSearchParams({
      user_id: params.user_id.toString(),
    });
    return apiRequest<Certification>(`/api/certifications?${queryParams.toString()}`, {
      method: "POST",
      body: JSON.stringify(params.data),
    });
  },

  /**
   * Update a certification.
   */
  update: (certificationId: number, data: CertificationUpdate): Promise<Certification> => {
    return apiRequest<Certification>(`/api/certifications/${certificationId}`, {
      method: "PATCH",
      body: JSON.stringify(data),
    });
  },

  /**
   * Delete a certification.
   */
  delete: (certificationId: number): Promise<void> => {
    return apiRequest<void>(`/api/certifications/${certificationId}`, {
      method: "DELETE",
    });
  },
};
